File: app/src/main/java/com/cicerodev/whatsapp/fragment/ConversasFragment.java
package com.cicerodev.whatsapp.fragment;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.cicerodev.whatsapp.R;
import com.cicerodev.whatsapp.adapter.ConversasAdapter;
import com.cicerodev.whatsapp.model.Conversa;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DatabaseReference;

import java.util.ArrayList;
import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 */
public class ConversasFragment extends Fragment {

    private RecyclerView recyclerViewConversas;
    private List<Conversa> listaConversas = new ArrayList<>();
    private ConversasAdapter adapter;
    private DatabaseReference database;
    private DatabaseReference conversasRef;
    private ChildEventListener childEventListenerConversas;

    public ConversasFragment() {}

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_conversas, container, false);

//        recyclerViewConversas = view.findViewById(R.id.recyclerListaConversas);
//
//        //Configurar adapter
//        adapter = new ConversasAdapter(listaConversas, getActivity());
//
//        //Configurar recyclerview
//        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getActivity());
//        recyclerViewConversas.setLayoutManager( layoutManager );
//        recyclerViewConversas.setHasFixedSize(true);
//        recyclerViewConversas.setAdapter( adapter );
//
//        //Configurar evento de clique
//        recyclerViewConversas.addOnItemTouchListener(
//                new RecyclerItemClickListener(
//                        getActivity(),
//                        recyclerViewConversas,
//                        new RecyclerItemClickListener.OnItemClickListener() {
//                            @Override
//                            public void onItemClick(View view, int position) {
//
//                                List<Conversa> listaConversasAtualizada = adapter.getConversas();
//                                Conversa conversaSelecionada = listaConversasAtualizada.get( position );
//
//                                if( conversaSelecionada.getIsGroup().equals("true") ){
//
//                                    Intent i = new Intent(getActivity(), ChatActivity.class);
//                                    i.putExtra("chatGrupo", conversaSelecionada.getGrupo() );
//                                    startActivity( i );
//
//                                }else {
//                                    Intent i = new Intent(getActivity(), ChatActivity.class);
//                                    i.putExtra("chatContato", conversaSelecionada.getUsuarioExibicao() );
//                                    startActivity( i );
//                                }
//
//                            }
//
//                            @Override
//                            public void onLongItemClick(View view, int position) {
//
//                            }
//
//                            @Override
//                            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
//
//                            }
//                        }
//                )
//        );
//
//
//        //Configura conversas ref
//        String identificadorUsuario = UsuarioFirebase.getIdentificadorUsuario();
//        database = ConfiguracaoFirebase.getFirebaseDatabase();
//        conversasRef = database.child("conversas")
//                .child( identificadorUsuario );

        return view;
    }

//    @Override
//    public void onStart() {
//        super.onStart();
//        recuperarConversas();
//    }
//
//    @Override
//    public void onStop() {
//        super.onStop();
//        conversasRef.removeEventListener( childEventListenerConversas );
//    }
//
//    public void recarregarConversas(){
//        adapter = new ConversasAdapter(listaConversas, getActivity());
//        recyclerViewConversas.setAdapter(adapter);
//        adapter.notifyDataSetChanged();
//    }
//
//    public void recuperarConversas(){
//
//        childEventListenerConversas = conversasRef.addChildEventListener(new ChildEventListener() {
//            @Override
//            public void onChildAdded(DataSnapshot dataSnapshot, String s) {
//
//                // Recuperar conversas
//                Conversa conversa = dataSnapshot.getValue( Conversa.class );
//                listaConversas.add( conversa );
//                adapter.notifyDataSetChanged();
//
//            }
//
//            @Override
//            public void onChildChanged(DataSnapshot dataSnapshot, String s) {
//
//            }
//
//            @Override
//            public void onChildRemoved(DataSnapshot dataSnapshot) {
//
//            }
//
//            @Override
//            public void onChildMoved(DataSnapshot dataSnapshot, String s) {
//
//            }
//
//            @Override
//            public void onCancelled(DatabaseError databaseError) {
//
//            }
//        });
//
//    }

}

File: app/src/main/java/com/cicerodev/whatsappcomdi/ui/fragments/conversas/ConversasFragment.kt
package com.cicerodev.whatsappcomdi.ui.fragments.conversas

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import androidx.fragment.app.viewModels
import androidx.navigation.navGraphViewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cicerodev.whatsappcomdi.R
import com.cicerodev.whatsappcomdi.adapter.ConversasAdapter
import com.cicerodev.whatsappcomdi.adapter.RecyclerItemClickListener
import com.cicerodev.whatsappcomdi.data.model.Conversa
import com.cicerodev.whatsappcomdi.databinding.FragmentConversasBinding
import com.cicerodev.whatsappcomdi.extensions.navigateTo
import com.cicerodev.whatsappcomdi.ui.base.BaseFragment
import com.cicerodev.whatsappcomdi.ui.fragments.home.HomeFragmentDirections
import com.cicerodev.whatsappcomdi.util.toast
import dagger.hilt.android.AndroidEntryPoint


@AndroidEntryPoint
class ConversasFragment : BaseFragment<FragmentConversasBinding, ConversasViewModel>() {

    private lateinit var adapter: ConversasAdapter
    override val viewModel: ConversasViewModel by viewModels()
    var listaConveras = mutableListOf<Conversa>()

    override fun getViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentConversasBinding = FragmentConversasBinding.inflate(inflater, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        configuraRecyclerViewListaConversas()
    }

    private fun configuraRecyclerViewListaConversas() {
        viewModel.conversa.observe(viewLifecycleOwner) { listaConversasLiveData ->
            //Configurar adapter
             adapter = ConversasAdapter(listaConversasLiveData, requireActivity())

            //Configurar recyclerview
            val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(requireContext())
            binding.recyclerListaConversas.setLayoutManager(layoutManager)
            binding.recyclerListaConversas.setHasFixedSize(true)
            binding.recyclerListaConversas.setAdapter(adapter)

            //Configurar evento de clique
            binding.recyclerListaConversas.addOnItemTouchListener(
                RecyclerItemClickListener(
                    activity,
                    binding.recyclerListaConversas,
                    object : RecyclerItemClickListener.OnItemClickListener {
                        override fun onItemClick(view: View?, position: Int) {
                            val listaConversasAtualizada = adapter.conversas
                            val conversaSelecionada = listaConversasAtualizada[position]
                            if (!conversaSelecionada.equals(null)) {
                                if (conversaSelecionada.isGroup == "true") {
                                    navigateTo(HomeFragmentDirections.actionHomeFragmentToChatFragment(null, "chatGrupo", conversaSelecionada.grupo))

                                } else {
                                    navigateTo(HomeFragmentDirections.actionHomeFragmentToChatFragment(conversaSelecionada.usuarioExibicao, "chatContato", null))
                                }
                            }
                        }

                        override fun onItemClick(
                            parent: AdapterView<*>?,
                            view: View?,
                            position: Int,
                            id: Long
                        ) {

                        }

                        override fun onLongItemClick(view: View?, position: Int) {

                        }

                    }
                )
            )
        }
    }

    fun pesquisarConversas(texto: String) {
        val listaConversasBusca = mutableListOf<Conversa>()
        for (conversa in viewModel.conversa.value!!) {
            val nome = conversa.usuarioExibicao?.nome?.lowercase()
            val ultimaMsg = conversa.ultimaMensagem.lowercase()

            if (nome?.contains(texto) == true || ultimaMsg.contains(texto)) {
               listaConversasBusca.add(conversa)

            }
        }
        adapter = ConversasAdapter(listaConversasBusca, requireContext())
        binding.recyclerListaConversas.adapter = adapter
        adapter.notifyDataSetChanged()
    }

    fun recuperarConversas() {
        listaConveras = viewModel.conversa.value!!
        adapter = ConversasAdapter(listaConveras, requireContext())
        binding.recyclerListaConversas.adapter = adapter
        adapter.notifyDataSetChanged()
    }

    override fun onStart() {
        super.onStart()
        viewModel.getConversas()
    }
}